Enhancement/2288 user input e2e tests

eugene-manuilov opened the pull request:
## Summary

<!-- Please reference the issue this PR addresses. -->
Addresses issue #2288

## Relevant technical choices

<!-- Please describe your changes. -->

## Checklist

- [x] My code is tested and passes existing unit tests.
- [ ] My code has an appropriate set of unit tests which all pass.
- [x] My code is backward-compatible with WordPress 4.7 and PHP 5.6.
- [x] My code follows the [WordPress](https://make.wordpress.org/core/handbook/best-practices/coding-standards/) coding standards.
- [x] My code has proper inline documentation.
- [ ] I have added a QA Brief on the issue linked above.
- [x] I have signed the Contributor License Agreement (see <https://cla.developers.google.com/>).

aaemnnosttv commented on tests/e2e/specs/user-input-questions.test.js:
```diff
@@ -0,0 +1,152 @@
+
+	async function expectSuccessNotification() {
+		await page.waitForSelector( '#user-input-success' );
+		await expect( page ).toMatchElement( '#user-input-success' );
```
This is redundant here as `waitForSelector` will fail the test if it times out (default 30s).

We shouldn't need to wait that long so I think it would be better to use just the `toMatchElement` here, but using the `timeout` option to use a more reasonable limit in the event of a legitimate failure (maybe 5s?).

eugene-manuilov replied:
This is needed because we should wait till the page is redirected. If I remove it, tests fail.

eugene-manuilov replied:
What we can remove is the expect clause: `await expect( page ).toMatchElement( '#user-input-success' );`. Updated.

aaemnnosttv commented on tests/e2e/plugins/user-input.php:
```diff
@@ -0,0 +1,139 @@
+	}
+
+	$user_id  = get_current_user_id();
+	$defaults = array(
```
Is all of this necessary/used by the tests? It looks like most of this might have been copied from before so I realize it probably wasn't all written for this, but it would be good to keep this to only what's needed if we can reduce it. What do you think?

eugene-manuilov replied:
It's needed because that's what we expect to receive even if there is nothing saved yet. We can't remove it.

aaemnnosttv commented on tests/e2e/plugins/user-input.php:
```diff
@@ -0,0 +1,139 @@
+add_filter( 'pre_http_request', 'e2e_user_input_settings_proxy_handler', 10, 3 );
+
+function e2e_user_input_settings_reset() {
+	$user_options     = new User_Options( new Context( GOOGLESITEKIT_PLUGIN_MAIN_FILE ) );
+	$user_input_state = new User_Input_State( $user_options );
```
While this works fine for the tests, we should account for this potentially being activated (or deactivated) with Site Kit inactive. Other E2E plugins check for a constant or use Site Kit provided actions to prevent such errors.

Maybe we should `wp_die` on `plugins_loaded` if Site Kit isn't active? I'm open to ideas here.

eugene-manuilov replied:
Updated to check whether the `GOOGLESITEKIT_PLUGIN_MAIN_FILE` constant is defined.

aaemnnosttv commented on assets/js/googlesitekit/datastore/user/user-input-settings.js:
```diff
@@ -109,7 +109,7 @@ const baseActions = {
 		yield clearError( 'saveUserInputSettings', [] );
 
-		const settings = registry.select( STORE_NAME ).getUserInputSettings();
+		const settings = registry.select( STORE_NAME ).getUserInputSettings() || {};
```
It seems like maybe a test should be added/updated for the changes to this action.

eugene-manuilov replied:
This seems to be not needed anymore. Reverted it.

aaemnnosttv reviewed (changes requested):
Thanks @eugene-manuilov – this worked well in my testing, and just a few things below to address, but nothing major; nice work!

One thing that I noticed (which I had to change to get the tests to run) was a PHP notice that seems to indicate a problem in the `User_Input_Settings` class.

```
Notice: Array to string conversion in /var/www/html/wp-content/plugins/google-site-kit/includes/Core/Util/User_Input_Settings.php on line 134 Warning: Cannot modify header information - headers already sent by (output started at /var/www/html/wp-content/plugins/google-site-kit/includes/Core/Util/User_Input_Settings.php:134) in /var/www/html/wp-admin/includes/misc.php on line 1310
Warning: Cannot modify header information - headers already sent by (output started at /var/www/html/wp-content/plugins/google-site-kit/includes/Core/Util/User_Input_Settings.php:134) in /var/www/html/wp-includes/pluggable.php on line 1296
Warning: Cannot modify header information - headers already sent by (output started at /var/www/html/wp-content/plugins/google-site-kit/includes/Core/Util/User_Input_Settings.php:134) in /var/www/html/wp-includes/pluggable.php on line 1299
```
Can you take a look?

Also, I noticed that when the success notice is shown, the CTA to provide user input is also showing:

![image](https://user-images.githubusercontent.com/1621608/106753954-0817a800-6635-11eb-8ba0-31eccd2d3784.png)

eugene-manuilov commented:
> One thing that I noticed (which I had to change to get the tests to run) was a PHP notice that seems to indicate a problem in the User_Input_Settings class.

This should be fixed.

> Also, I noticed that when the success notice is shown, the CTA to provide user input is also showing:

This happens because we don't use a real proxy server and the authentication class can't set the appropriate user input flag.

eugene-manuilov commented:
@aaemnnosttv this is ready for the next round of review. I have reworked it slightly, plus added a few utility functions and updated e2e actions to help with troubleshooting. Hope it works for you.

aaemnnosttv commented on includes/Core/Util/User_Input_Settings.php:
```diff
@@ -128,18 +128,22 @@ function( $setting ) {
 			),
 		);
 
+		if ( ! empty( $access_token ) && is_string( $access_token ) ) {
+			$user_input_settings_args['headers']['Authorization'] = 'Bearer ' . $access_token;
+		}
+
```
Why change this? The access token is required for making this request, is it not?

eugene-manuilov replied:
Had to change it because we don't use real OAuth client during e2e tests, thus it doesn't return an access token and causes unexpected issues.

> The access token is required for making this request, is it not?

Yes, it is required by the proxy server itself. The user input settings class just passes what it has.

aaemnnosttv replied:
Hmm, okay. I suppose this is fine then since we weren't guarding against making the request if there was no token before.

aaemnnosttv commented on tests/e2e/utils/step-and-screenshot.js:
```diff
@@ -0,0 +1,78 @@
+ * @return {Promise} Promise object.
+ */
+export function step( name, cb ) {
+	return new Promise( async ( success, error ) => {
```
Minor difference, but let's use the same names we do elsewhere for consistency.
```suggestion
	return new Promise( async ( resolve, reject ) => {
```

aaemnnosttv commented on tests/e2e/utils/step-and-screenshot.js:
```diff
@@ -0,0 +1,78 @@
+		try {
+			const results = await ( typeof cb === 'function' ? cb() : cb );
+			await screenshot( name );
+			success( results );
```
```suggestion
			resolve( results );
```

aaemnnosttv commented on tests/e2e/utils/step-and-screenshot.js:
```diff
@@ -0,0 +1,78 @@
+			success( results );
+		} catch ( err ) {
+			await screenshot( name );
+			error( err );
```
```suggestion
			reject( err );
```

aaemnnosttv commented on tests/e2e/utils/step-and-screenshot.js:
```diff
@@ -0,0 +1,78 @@
+	await page.screenshot( {
+		path: `${ screenshotsDir }/${ screenshotIndex.toString().padStart( 2, '0' ) }-${ name.replace( /\W+/, '-' ) }.png`,
+		type: 'png',
+		fullPage: true,
```
I would prefer we don't default to `fullPage` here as this loses what the browser "sees" at that moment in the viewport which can be relevant as to why the test failed. I see how this could be useful in some cases though so perhaps we should accept an optional second argument for screenshot options that we can spread into the call here. We could then add this as the 3rd parameter to `step` below.

eugene-manuilov replied:
Updated.

aaemnnosttv reviewed (commented):
Thanks @eugene-manuilov – overall this is looking good and close to ready. Just a few questions and suggestions from me.

aaemnnosttv commented on tests/e2e/utils/step-and-screenshot.js:
```diff
@@ -0,0 +1,78 @@
+ * @return {Promise} Promise object.
+ */
+export function step( name, cb ) {
+	return new Promise( async ( resolve, reject ) => {
+		try {
```
Regarding the lint errors, let's import `path from path` rather than using named imports for those functions to avoid this conflict.

eugene-manuilov replied:
Yep, already done it on my end.

aaemnnosttv reviewed (approved):
LGTM 👍  Nice touch with the added screenshot artifacts!
